MessageQueue/ServerStatisticsReceiver.cs

```diff
@@ -0,0 +1,104 @@
+﻿using System.Text;
+using System.Text.Json;
+using RabbitMQ.Client;
+using RabbitMQ.Client.Events;
+using ServerStatistics.Extensions;
+using ServerStatistics.Models;
+using ServerStatistics.Services;
+using SignalREndpoint;
+
+namespace MessageQueue
+{
+    public class ServerStatisticsReceiver : RabbitMqConnector, IMessageReceiver
+    {
+        private string _queueName;
+        private IServerStatisticsService _statisticsService;
+        private ServerStatisticsDTO _serverStatistics;
+        private IAlertSender _alertSender;
+        private bool _hasPreviousStatistics = false;
+        private ServerStatisticsDTO _previousStatistics;
+
+        public ServerStatisticsReceiver(
+            string connectionName,
+            string exchangeName,
+            string queueName,
+            string routingKey,
+            IServerStatisticsService statisticsService,
+            IAlertSender alertSender) : base(connectionName)
+        {
+            _channel.ExchangeDeclare(exchangeName, ExchangeType.Topic);
+            _channel.QueueDeclare(queueName, false, false, false);
+            _channel.QueueBind(queueName, exchangeName, routingKey);
+            _queueName = queueName;
+            _statisticsService = statisticsService;
+            _alertSender = alertSender;
+        }
+
+        public void Consume()
+        {
+            var consumer = new EventingBasicConsumer(_channel);
+
+            consumer.Received += async (sender, args) =>
+            {
+                try
+                {
+                    _serverStatistics = GetServerStatistics(args);
+
+                    await DetectAndReportAnomaliesAsync();
+                    await PersistToDatabaseAsync(args);
+
+                    _channel.BasicAck(args.DeliveryTag, false);
+                }
+                catch { }
+            };
+
+            _channel.BasicConsume(_queueName, false, consumer);
+        }
+
+        private ServerStatisticsDTO? GetServerStatistics(BasicDeliverEventArgs args)
+        {
+            var bytesMessage = args.Body.ToArray();
+            var stringMessage = Encoding.UTF8.GetString(bytesMessage);
+
+            return JsonSerializer.Deserialize<ServerStatisticsDTO>(stringMessage);
+        }
+
+        private async Task PersistToDatabaseAsync(BasicDeliverEventArgs args)
+        {
+            await _statisticsService.InsertAsync(GetServerStatisticsWithServerIdentifier(args));
+        }
+
+        private ServerStatisticsWithServerIdentifierDTO GetServerStatisticsWithServerIdentifier(
+            BasicDeliverEventArgs args)
+        {
+            var serverIdentifier = args.RoutingKey
+                .Split('.')
+                .LastOrDefault();
+
+            return _serverStatistics.IncludeServerIdentifier(serverIdentifier);
+        }
+
+        private async Task DetectAndReportAnomaliesAsync()
```

Comment:
I think both of these methods should be moved to a separate service, this receiver doesn't have to know that much of details.

Reply:
I see that you maintain `_previousStatistics` in this class, but still feels like this logic needs to be moved out